clients/src/main/java/org/apache/kafka/common/requests/FetchResponse.java

```diff
@@ -59,136 +60,140 @@
  */
 public class FetchResponse<T extends BaseRecords> extends AbstractResponse {
 
-    public static final long INVALID_HIGHWATERMARK = -1L;
+    public static FetchResponseData.FetchablePartitionResponse partitionResponse(Errors error) {
+        return new FetchResponseData.FetchablePartitionResponse()
+                .setErrorCode(error.code())
+                .setHighWatermark(FetchResponse.INVALID_HIGH_WATERMARK)
+                .setLastStableOffset(FetchResponse.INVALID_LAST_STABLE_OFFSET)
+                .setLogStartOffset(FetchResponse.INVALID_LOG_START_OFFSET)
+                .setAbortedTransactions(null)
+                .setRecordSet(MemoryRecords.EMPTY)
+                .setPreferredReadReplica(FetchResponse.INVALID_PREFERRED_REPLICA_ID);
+    }
+
+    public static final long INVALID_HIGH_WATERMARK = -1L;
     public static final long INVALID_LAST_STABLE_OFFSET = -1L;
     public static final long INVALID_LOG_START_OFFSET = -1L;
     public static final int INVALID_PREFERRED_REPLICA_ID = -1;
 
     private final FetchResponseData data;
-    private final LinkedHashMap<TopicPartition, PartitionData<T>> responseDataMap;
+    private final LinkedHashMap<TopicPartition, FetchResponseData.FetchablePartitionResponse> partitionData;
+    // lazily instantiate this field
+    private LinkedHashMap<TopicPartition, PartitionData<T>> responseDataMap = null;
```

Comment:
Can we remove this altogether?

Reply:
```partitionData``` can be removed but ```responseDataMap``` is still a response type and it is used by production code. I did not refactor that to avoid big patch.

Reply:
I understand. It's a bit hard to understand the end state for this class as it stands. We've taken a few intermediate steps to this point. Can we just go to the end state now?

Reply:
You are right. Will address it in next commit.